services/streammanager/kafka/client/producer.go

```diff
@@ -0,0 +1,120 @@
+package client
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/segmentio/kafka-go"
+)
+
+type ProducerConfig struct {
+	ClientID string
+	WriteTimeout,
+	ReadTimeout time.Duration
+	Logger      Logger
+	ErrorLogger Logger
+}
+
+func (c *ProducerConfig) defaults() {
+	if c.WriteTimeout < 1 {
+		c.WriteTimeout = 10 * time.Second
+	}
+	if c.ReadTimeout < 1 {
+		c.ReadTimeout = 10 * time.Second
+	}
+}
+
+// Producer provides a high-level API for producing messages to Kafka
+type Producer struct {
+	writer *kafka.Writer
+}
+
+// NewProducer instantiates a new producer. To use it asynchronously just do "go p.Publish(ctx, msgs)".
+func (c *Client) NewProducer(topic string, producerConf ProducerConfig) (p *Producer, err error) {
+	producerConf.defaults()
+
+	dialer := &net.Dialer{
+		Timeout: c.config.DialTimeout,
+	}
+	transport := &kafka.Transport{
+		DialTimeout: c.config.DialTimeout,
+		Dial:        dialer.DialContext,
+	}
+	if producerConf.ClientID != "" {
+		transport.ClientID = producerConf.ClientID
+	} else if c.config.ClientID != "" {
+		transport.ClientID = c.config.ClientID
+	}
+	if c.config.TLS != nil {
+		transport.TLS, err = c.config.TLS.build()
+		if err != nil {
+			return nil, fmt.Errorf("could not build TLS configuration: %w", err)
+		}
+	}
+	if c.config.SASL != nil {
+		transport.SASL, err = c.config.SASL.build()
+		if err != nil {
+			return nil, fmt.Errorf("could not build SASL configuration: %w", err)
+		}
+	}
+
+	p = &Producer{
+		writer: &kafka.Writer{
```

Comment:
Judging by [this](https://github.com/segmentio/kafka-go/blob/main/writer.go#L796), I believe we should explicitly set `BatchTimeout` to something very small, e.g. `1*time.NanoSecond`.

Reply:
I thought it had to be greater than one to create issues. The default is `1` though and while testing I didn't encounter any delays, it seems to send messages synchronously.

> // When sending synchronously and the writer's batch size is configured to be
> // greater than 1, this method blocks until either a full batch can be assembled
> // or the batch timeout is reached.

But if we don't provide it then it defaults to 1s:

```go
func (w *Writer) batchTimeout() time.Duration {
	if w.BatchTimeout > 0 {
		return w.BatchTimeout
	}
	return 1 * time.Second
}
```

* https://github.com/segmentio/kafka-go/blob/main/writer.go#L796
* https://github.com/segmentio/kafka-go/blob/6691dda207ff50d6a50baed99748c620d569aaa9/writer.go#L577

It would be very weird that in order to make it sync people would have to come up with the 1µs setting btw, so it seems OK to me :thinking: 

WDYT?

Reply:
The default BatchSize is [100](https://github.com/segmentio/kafka-go/blob/6691dda207ff50d6a50baed99748c620d569aaa9/writer.go#L782), so according to the documentation it will wait. Setting the `BatchSize` to 0 might make more sense, as long as this doesn't have as a side effect to send a varArg of messages in separate requests, which I don't think so.

Reply:
What the hell I confused size with timeout :laughing: Thanks man, I'll look into it then :+1: